src/google-classroom-extractor/edfi_google_classroom_extractor/mapping/assignment_submissions.py

```diff
@@ -63,6 +66,39 @@ def derive_state(submission_row: Series) -> str:
     return api_state
 
 
+def _get_submission_datetime(submission_row: Series) -> str:
+    if (
+        submission_row["state"] == TURNED_IN_STATE
+        or submission_row["state"] == RETURNED_STATE
+    ):
+        return submission_row["updateTime"]
+
+    status_history: List[dict] = ast.literal_eval(submission_row["submissionHistory"])
+    # this submission history also has a history of grades, but for this purpose, we only care
+    # about stateHistory so we filter by stateHistory
+    status_history = [
+        status_item["stateHistory"]
+        for status_item in status_history
+        if "stateHistory" in status_item.keys()
+    ]
+    status_history = sorted(
+        status_history, key=lambda k: k["stateTimestamp"], reverse=True
+    )
+
+    for status_record in status_history:
+        status_is_turned_in: bool = status_record["state"] == TURNED_IN_STATE
+        status_is_returned: bool = status_record["state"] == RETURNED_STATE
+        status_is_reclaimed_by_student: bool = status_record["state"] == RECLAIMED_STATE
+
+        if status_is_reclaimed_by_student:
+            return ""
+
+        if status_is_turned_in or status_is_returned:
+            return status_record["stateTimestamp"]
+
+    return ""
```

Comment:
Some experimental code... not sure if this is any better, just different. I just needed some practice in Python!

```python
df = pd.DataFrame(
    [r["stateHistory"] for r in submission_row["submissionHistory"] if "stateHistory" in r]
)

df.sort_values(by=["stateTimestamp"], inplace=True,  ascending=False)

return df[df["state"] == "TURNED_IN"].iloc[0]["stateTimestamp"]
```